File: sdk/cpp/lib/ml/rl/cartPole.cpp
#include <ml/rl/cartPole.h>

#include <oa/core/bufferAccess.h>
#include <oa/core/fnMatrix.h>
#include <oa/core/op.h>
#include <oa/runtime/executionSession.h>
#include <oa/runtime/engine.h>

#include <ml/rl/environmentKernelPack.h>
#include <ml/rl/gen/environmentOpRegistry.h>
#include <oa/ml/environmentExecution.h>

#include <bit>
#include <cmath>
#include <limits>

namespace {

constexpr oa::U64 FnvOffset = 14695981039346656037ULL;
constexpr oa::U64 FnvPrime = 1099511628211ULL;

void hashU32(oa::U64& inOutHash, oa::U32 inValue) noexcept {
	for (oa::U32 shift = 0; shift < 32U; shift += 8U) {
		inOutHash ^= static_cast<oa::U8>(inValue >> shift);
		inOutHash *= FnvPrime;
	}
}

} // namespace

namespace oa {

oa::U64 CartPoleConfig::dynamicsIdentity() const noexcept {
	oa::U64 hash = FnvOffset;
	hashU32(hash, dynamicsVersion);
	hashU32(hash, maxEpisodeSteps);
	hashU32(hash, std::bit_cast<oa::U32>(gravity));
	hashU32(hash, std::bit_cast<oa::U32>(cartMass));
	hashU32(hash, std::bit_cast<oa::U32>(poleMass));
	hashU32(hash, std::bit_cast<oa::U32>(halfPoleLength));
	hashU32(hash, std::bit_cast<oa::U32>(forceMagnitude));
	hashU32(hash, std::bit_cast<oa::U32>(timeStep));
	hashU32(hash, std::bit_cast<oa::U32>(positionThreshold));
	hashU32(hash, std::bit_cast<oa::U32>(angleThresholdRadians));
	return hash;
}

bool CartPoleStep::isValid() const noexcept {
	return !observation.isEmpty() && !nextObservation.isEmpty()
		&& !reward.isEmpty() && !terminated.isEmpty()
		&& !truncated.isEmpty() && !done.isEmpty();
}

CartPole::CartPole(oa::Engine& inEngine)
	: oa::Environment(inEngine) {}

oa::U64 CartPole::effectiveSeed_() const noexcept {
	return hasPendingSeed_ ? pendingSeed_ : config_.seed;
}

oa::Result<CartPole> CartPole::create(
	oa::Engine& inEngine,
	const CartPoleConfig& inConfig) {
	const bool finite = std::isfinite(inConfig.gravity)
		&& std::isfinite(inConfig.cartMass)
		&& std::isfinite(inConfig.poleMass)
		&& std::isfinite(inConfig.halfPoleLength)
		&& std::isfinite(inConfig.forceMagnitude)
		&& std::isfinite(inConfig.timeStep)
		&& std::isfinite(inConfig.positionThreshold)
		&& std::isfinite(inConfig.angleThresholdRadians);
	if (!finite || inConfig.environments == 0 || inConfig.maxEpisodeSteps == 0
		|| inConfig.cartMass <= 0.0F || inConfig.poleMass <= 0.0F
		|| inConfig.halfPoleLength <= 0.0F || inConfig.forceMagnitude <= 0.0F
		|| inConfig.timeStep <= 0.0F || inConfig.positionThreshold <= 0.0F
		|| inConfig.angleThresholdRadians <= 0.0F) {
		return oa::Status::invalidArgument(
			"CartPole::create received an invalid environment configuration");
	}
	if (inConfig.environments > std::numeric_limits<oa::U32>::max() / 4U) {
		return oa::Status::error(
			oa::StatusCode::OutOfRange,
			"CartPole::create exceeds the 32-bit GPU indexing limit");
	}
	if (!inEngine.isReady()) {
		return oa::Status::error(
			oa::StatusCode::FailedPrecondition,
			"CartPole::create requires a ready engine");
	}
	OA_RETURN_IF_ERROR(oa::ensureEnvironmentKernelPack(inEngine));

	const oa::MatrixShape stateShape{
		static_cast<oa::I64>(inConfig.environments), 4};
	const oa::MatrixShape vectorShape{
		static_cast<oa::I64>(inConfig.environments)};
	CartPole result(inEngine);
	if (!result.isOpen()) {
		return oa::Status::error(
			oa::StatusCode::FailedPrecondition,
			"CartPole::create could not open its execution session");
	}
	result.config_ = inConfig;
	result.spec_ = {
		.observation = oa::EnvironmentSpace::box(
			"observation", {4}, oa::ScalarType::Float32),
		.action = oa::EnvironmentSpace::discrete("action", 2),
		.reward = oa::EnvironmentSpace::box(
			"reward", {}, oa::ScalarType::Float32, 0.0, 1.0),
		.terminated = oa::EnvironmentSpace::binary("terminated"),
		.truncated = oa::EnvironmentSpace::binary("truncated"),
	};
	OA_RETURN_IF_ERROR(result.spec_.validateDefinition());
	const oa::Status initialized = result.recordCommands([&]() -> oa::Status {
		result.state_ = oa::FnMatrix::empty(stateShape, oa::ScalarType::Float32);
		result.transitionObservation_ = oa::FnMatrix::empty(
			stateShape, oa::ScalarType::Float32);
		result.reward_ = oa::FnMatrix::empty(vectorShape, oa::ScalarType::Float32);
		result.terminated_ = oa::FnMatrix::empty(vectorShape, oa::ScalarType::UInt8);
		result.truncated_ = oa::FnMatrix::empty(vectorShape, oa::ScalarType::UInt8);
		result.done_ = oa::FnMatrix::empty(vectorShape, oa::ScalarType::UInt8);
		result.episodeSteps_ = oa::FnMatrix::empty(
			vectorShape, oa::ScalarType::UInt32);
		result.episodeIndex_ = oa::FnMatrix::empty(
			vectorShape, oa::ScalarType::UInt32);
		if (!result.isValid()) {
			return oa::Status::error(
				oa::StatusCode::OutOfMemory,
				"CartPole::create could not allocate environment storage");
		}
		return result.recordReset_(false);
	});
	if (initialized.isError()) return initialized;
	return result;
}

bool CartPole::isValid() const noexcept {
	const oa::MatrixShape stateShape{
		static_cast<oa::I64>(config_.environments), 4};
	const oa::MatrixShape vectorShape{
		static_cast<oa::I64>(config_.environments)};
	const auto matches = [](const oa::Matrix& inMatrix,
		const oa::MatrixShape& inShape, oa::ScalarType inDtype) {
		return !inMatrix.isEmpty() && inMatrix.getShape() == inShape
			&& inMatrix.getDtype() == inDtype;
	};
	return matches(state_, stateShape, oa::ScalarType::Float32)
		&& matches(transitionObservation_, stateShape, oa::ScalarType::Float32)
		&& matches(reward_, vectorShape, oa::ScalarType::Float32)
		&& matches(terminated_, vectorShape, oa::ScalarType::UInt8)
		&& matches(truncated_, vectorShape, oa::ScalarType::UInt8)
		&& matches(done_, vectorShape, oa::ScalarType::UInt8)
		&& matches(episodeSteps_, vectorShape, oa::ScalarType::UInt32)
		&& matches(episodeIndex_, vectorShape, oa::ScalarType::UInt32);
}

oa::Status CartPole::recordReset_(bool inOnlyDone) {
	if (!isValid()) return oa::Status::error(
		oa::StatusCode::FailedPrecondition,
			"CartPole reset requires a valid environment");
	if (inOnlyDone && !hasCommittedState_ && !hasPendingFullReset_) {
		return oa::Status::error(
			oa::StatusCode::FailedPrecondition,
			"CartPole completed reset requires submitted state or an earlier full reset in this transaction");
	}
	const oa::U64 seed = effectiveSeed_();
	if (!inOnlyDone) hasPendingFullReset_ = true;
	struct Push {
		oa::U32 environments;
		oa::U32 seedLow;
		oa::U32 seedHigh;
		oa::U32 onlyDone;
	} push{
		.environments = config_.environments,
		.seedLow = static_cast<oa::U32>(seed),
		.seedHigh = static_cast<oa::U32>(seed >> 32U),
		.onlyDone = inOnlyDone ? 1U : 0U,
	};
	oa::BufferAccess access[] = {
		oa::BufferAccess::ReadWrite, oa::BufferAccess::Write,
		oa::BufferAccess::Write, oa::BufferAccess::ReadWrite,
	};
	auto& context = oa::EnvironmentExecutionAccess::session(*this);
	const auto semantic = context.recordOp(
		oa::detail::opRegistry::FnEnvironment::cartPoleReset,
		{&done_, &state_, &episodeSteps_, &episodeIndex_},
		{&done_, &state_, &episodeSteps_, &episodeIndex_},
		{
			oa::OpAttribute::fromUnsignedInteger("seed", seed),
			oa::OpAttribute::fromBoolean("onlyCompleted", inOnlyDone),
		});
	if (semantic.isError()) return semantic.getStatus();
	context.add(
		"RlCartPoleReset",
		{&done_, &state_, &episodeSteps_, &episodeIndex_},
		access, &push, sizeof(push),
		(config_.environments + 255U) / 256U, 1, 1,
		oa::detail::opRegistry::FnEnvironment::cartPoleReset.name, 0,
		oa::detail::opRegistry::FnEnvironment::cartPoleReset.hash, 0, 0,
		semantic.getValue());
	return oa::Status::ok();
}

oa::Status CartPole::reset() {
	return oa::Environment::reset(effectiveSeed_());
}

oa::Status CartPole::resetDone() {
	return resetCompleted();
}

oa::Status CartPole::recordReset_(oa::U64 inSeed) {
	pendingSeed_ = inSeed;
	hasPendingSeed_ = true;
	return recordReset_(false);
}

oa::Status CartPole::recordResetCompleted_() {
	return recordReset_(true);
}

oa::Result<oa::EnvironmentTransition> CartPole::recordStep_(
	const oa::Matrix& inAction) {
	auto step = recordDetailedStep_(inAction);
	if (step.isError()) return step.getStatus();
	return oa::EnvironmentTransition{
		.observation = step->observation,
		.nextObservation = step->nextObservation,
		.reward = step->reward,
		.terminated = step->terminated,
		.truncated = step->truncated,
	};
}

oa::Result<CartPoleStep> CartPole::step(
	const oa::Matrix& inAction) {
	auto transition = oa::Environment::step(inAction);
	if (transition.isError()) return transition.getStatus();
	return CartPoleStep{
		.observation = transition->observation,
		.nextObservation = transition->nextObservation,
		.reward = transition->reward,
		.terminated = transition->terminated,
		.truncated = transition->truncated,
		.done = done_,
	};
}

oa::Result<CartPoleStep> CartPole::recordDetailedStep_(
	const oa::Matrix& inAction) {
	if (!isValid()) {
		return oa::Status::error(
			oa::StatusCode::FailedPrecondition,
			"CartPole::step requires a valid environment");
	}
	if (!hasCommittedState_ && !hasPendingFullReset_) {
		return oa::Status::error(
			oa::StatusCode::FailedPrecondition,
			"CartPole::step requires submitted state or an earlier full reset in this transaction");
	}
	OA_RETURN_IF_ERROR(spec_.validateAction(inAction, config_.environments));

	struct Push {
		oa::U32 environments;
		oa::U32 maxEpisodeSteps;
		oa::F32 gravity;
		oa::F32 cartMass;
		oa::F32 poleMass;
		oa::F32 halfPoleLength;
		oa::F32 forceMagnitude;
		oa::F32 timeStep;
		oa::F32 positionThreshold;
		oa::F32 angleThresholdRadians;
	} push{
		.environments = config_.environments,
		.maxEpisodeSteps = config_.maxEpisodeSteps,
		.gravity = config_.gravity,
		.cartMass = config_.cartMass,
		.poleMass = config_.poleMass,
		.halfPoleLength = config_.halfPoleLength,
		.forceMagnitude = config_.forceMagnitude,
		.timeStep = config_.timeStep,
		.positionThreshold = config_.positionThreshold,
		.angleThresholdRadians = config_.angleThresholdRadians,
	};
	oa::BufferAccess access[] = {
		oa::BufferAccess::Read, oa::BufferAccess::ReadWrite,
		oa::BufferAccess::Write, oa::BufferAccess::Write,
		oa::BufferAccess::Write, oa::BufferAccess::Write,
		oa::BufferAccess::ReadWrite, oa::BufferAccess::ReadWrite,
	};
	auto& context = oa::EnvironmentExecutionAccess::session(*this);
	const auto semantic = context.recordOp(
		oa::detail::opRegistry::FnEnvironment::cartPoleStep,
		{&inAction, &state_, &done_, &episodeSteps_},
		{&transitionObservation_, &state_, &reward_, &terminated_,
		 &truncated_, &done_, &episodeSteps_},
		{
			oa::OpAttribute::fromUnsignedInteger(
				"dynamicsVersion", CartPoleConfig::dynamicsVersion),
			oa::OpAttribute::fromUnsignedInteger(
				"dynamicsIdentity", config_.dynamicsIdentity()),
			oa::OpAttribute::fromUnsignedInteger(
				"maxEpisodeSteps", config_.maxEpisodeSteps),
			oa::OpAttribute::fromFloat("gravity", config_.gravity),
			oa::OpAttribute::fromFloat(
				"forceMagnitude", config_.forceMagnitude),
			oa::OpAttribute::fromFloat("timeStep", config_.timeStep),
			oa::OpAttribute::fromFloat(
				"positionThreshold", config_.positionThreshold),
			oa::OpAttribute::fromFloat(
				"angleThresholdRadians", config_.angleThresholdRadians),
		});
	if (semantic.isError()) return semantic.getStatus();
	context.add(
		"RlCartPoleStep",
		{&inAction, &state_, &transitionObservation_, &reward_,
		 &terminated_, &truncated_, &done_, &episodeSteps_},
		access, &push, sizeof(push),
		(config_.environments + 255U) / 256U, 1, 1,
		oa::detail::opRegistry::FnEnvironment::cartPoleStep.name, 0,
		oa::detail::opRegistry::FnEnvironment::cartPoleStep.hash, 0, 0,
		semantic.getValue());
	return CartPoleStep{
		.observation = transitionObservation_,
		.nextObservation = state_,
		.reward = reward_,
		.terminated = terminated_,
		.truncated = truncated_,
		.done = done_,
	};
}

void CartPole::commitRecordedState_() noexcept {
	if (hasPendingSeed_) config_.seed = pendingSeed_;
	if (hasPendingFullReset_) hasCommittedState_ = true;
	hasPendingSeed_ = false;
	hasPendingFullReset_ = false;
}

void CartPole::rollbackRecordedState_() noexcept {
	hasPendingSeed_ = false;
	hasPendingFullReset_ = false;
}

} // namespace oa

File: sdk/rs/ml/rl/cart_pole.rs
//! SDK-owned vectorized CartPole-v1 environment.

use crate::ml::environment::{Environment, EnvironmentExecution};
use crate::ml::lowering::environment as lowering;
use crate::ml::{EnvironmentSpace, EnvironmentSpec, EnvironmentTransition};
use crate::{DType, Engine, Error, Matrix, Result};

/// Versioned dynamics and vector-lane policy for CartPole-v1.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct CartPoleConfig {
	/// Number of independent GPU environment lanes.
	pub environments: u32,
	/// Time-limit truncation threshold per episode.
	pub max_episode_steps: u32,
	/// Deterministic reset seed.
	pub seed: u64,
	/// Gravitational acceleration.
	pub gravity: f32,
	/// Cart mass.
	pub cart_mass: f32,
	/// Pole mass.
	pub pole_mass: f32,
	/// Half pole length used by the standard equations.
	pub half_pole_length: f32,
	/// Magnitude of the left/right action force.
	pub force_magnitude: f32,
	/// Euler integration interval.
	pub time_step: f32,
	/// Absolute cart-position termination threshold.
	pub position_threshold: f32,
	/// Absolute pole-angle termination threshold in radians.
	pub angle_threshold_radians: f32,
}

impl CartPoleConfig {
	/// Version of the complete step-dynamics identity.
	pub const DYNAMICS_VERSION: u32 = 1;

	/// Return a stable identity over every behavior-changing dynamics field.
	pub fn dynamics_identity(self) -> u64 {
		const OFFSET: u64 = 14_695_981_039_346_656_037;
		const PRIME: u64 = 1_099_511_628_211;
		fn hash_u32(mut hash: u64, value: u32) -> u64 {
			for byte in value.to_le_bytes() {
				hash ^= u64::from(byte);
				hash = hash.wrapping_mul(PRIME);
			}
			hash
		}
		[
			Self::DYNAMICS_VERSION,
			self.max_episode_steps,
			self.gravity.to_bits(),
			self.cart_mass.to_bits(),
			self.pole_mass.to_bits(),
			self.half_pole_length.to_bits(),
			self.force_magnitude.to_bits(),
			self.time_step.to_bits(),
			self.position_threshold.to_bits(),
			self.angle_threshold_radians.to_bits(),
		]
		.into_iter()
		.fold(OFFSET, hash_u32)
	}
}

impl Default for CartPoleConfig {
	fn default() -> Self {
		Self {
			environments: 1,
			max_episode_steps: 500,
			seed: 1,
			gravity: 9.8,
			cart_mass: 1.0,
			pole_mass: 0.1,
			half_pole_length: 0.5,
			force_magnitude: 10.0,
			time_step: 0.02,
			position_threshold: 2.4,
			angle_threshold_radians: 0.209_439_52,
		}
	}
}

/// Vectorized GPU CartPole workload over the reusable Environment lifecycle.
#[must_use]
pub struct CartPole<'engine> {
	engine: &'engine Engine,
	execution: EnvironmentExecution,
	config: CartPoleConfig,
	spec: EnvironmentSpec,
	state: Matrix,
	transition_observation: Matrix,
	reward: Matrix,
	terminated: Matrix,
	truncated: Matrix,
	done: Matrix,
	episode_steps: Matrix,
	episode_index: Matrix,
	pending_seed: Option<u64>,
	has_committed_state: bool,
	has_pending_full_reset: bool,
}

impl<'engine> CartPole<'engine> {
	/// Construct CartPole storage and record its deterministic initial reset.
	///
	/// # Errors
	///
	/// Returns an error for invalid dynamics, allocation, schema, or recording.
	pub fn new(engine: &'engine Engine, config: CartPoleConfig) -> Result<Self> {
		validate_config(config)?;
		let environments = config.environments as usize;
		let state = Matrix::from_f32(engine, [environments, 4], &vec![0.0; environments * 4])?;
		let transition_observation =
			Matrix::from_f32(engine, [environments, 4], &vec![0.0; environments * 4])?;
		let reward = Matrix::from_f32(engine, [environments], &vec![0.0; environments])?;
		let terminated = Matrix::from_slice(engine, [environments], &vec![0_u8; environments])?;
		let truncated = Matrix::from_slice(engine, [environments], &vec![0_u8; environments])?;
		let done = Matrix::from_slice(engine, [environments], &vec![0_u8; environments])?;
		let episode_steps = Matrix::from_slice(engine, [environments], &vec![0_u32; environments])?;
		let episode_index = Matrix::from_slice(engine, [environments], &vec![0_u32; environments])?;
		let spec = EnvironmentSpec::new(
			EnvironmentSpace::continuous(
				"observation",
				[4],
				DType::F32,
				f64::NEG_INFINITY,
				f64::INFINITY,
			)?,
			EnvironmentSpace::discrete("action", 2, DType::I32)?,
			EnvironmentSpace::continuous("reward", [], DType::F32, 0.0, 1.0)?,
			EnvironmentSpace::binary("terminated", [])?,
			EnvironmentSpace::binary("truncated", [])?,
		)?;
		let mut result = Self {
			engine,
			execution: EnvironmentExecution::new(engine),
			config,
			spec,
			state,
			transition_observation,
			reward,
			terminated,
			truncated,
			done,
			episode_steps,
			episode_index,
			pending_seed: None,
			has_committed_state: false,
			has_pending_full_reset: false,
		};
		Environment::reset(&mut result, config.seed)?;
		Ok(result)
	}

	/// Return the complete dynamics configuration.
	pub const fn config(&self) -> CartPoleConfig {
		self.config
	}

	/// Return the done mask used by completed-lane reset.
	pub const fn done(&self) -> &Matrix {
		&self.done
	}

	/// Return per-lane steps in the current episode.
	pub const fn episode_steps(&self) -> &Matrix {
		&self.episode_steps
	}

	/// Return per-lane deterministic episode indices.
	pub const fn episode_index(&self) -> &Matrix {
		&self.episode_index
	}

	fn effective_seed(&self) -> u64 {
		self.pending_seed.unwrap_or(self.config.seed)
	}

	fn record_reset_impl(&mut self, only_done: bool) -> Result<()> {
		if only_done && !self.has_committed_state && !self.has_pending_full_reset {
			return Err(Error::failed_precondition(
				"CartPole completed reset requires submitted state or an earlier full reset",
			));
		}
		if !only_done {
			self.has_pending_full_reset = true;
		}
		lowering::cart_pole_reset(
			&self.done,
			&self.state,
			&self.episode_steps,
			&self.episode_index,
			self.config.environments,
			self.effective_seed(),
			only_done,
		)
	}
}

impl Environment for CartPole<'_> {
	fn engine(&self) -> &Engine {
		self.engine
	}

	fn spec(&self) -> &EnvironmentSpec {
		&self.spec
	}

	fn environments(&self) -> u32 {
		self.config.environments
	}

	fn observation(&self) -> &Matrix {
		&self.state
	}

	fn execution(&self) -> &EnvironmentExecution {
		&self.execution
	}

	fn execution_mut(&mut self) -> &mut EnvironmentExecution {
		&mut self.execution
	}

	fn record_reset(&mut self, seed: u64) -> Result<()> {
		self.pending_seed = Some(seed);
		self.record_reset_impl(false)
	}

	fn record_step(&mut self, action: &Matrix) -> Result<EnvironmentTransition> {
		if !self.has_committed_state && !self.has_pending_full_reset {
			return Err(Error::failed_precondition(
				"CartPole step requires submitted state or an earlier full reset",
			));
		}
		lowering::cart_pole_step(
			action,
			&self.state,
			&self.transition_observation,
			&self.reward,
			&self.terminated,
			&self.truncated,
			&self.done,
			&self.episode_steps,
			self.config.environments,
			self.config.max_episode_steps,
			self.config.gravity,
			self.config.cart_mass,
			self.config.pole_mass,
			self.config.half_pole_length,
			self.config.force_magnitude,
			self.config.time_step,
			self.config.position_threshold,
			self.config.angle_threshold_radians,
			self.config.dynamics_identity(),
		)?;
		Ok(EnvironmentTransition::new(
			self.transition_observation.clone(),
			self.state.clone(),
			self.reward.clone(),
			self.terminated.clone(),
			self.truncated.clone(),
		))
	}

	fn record_reset_completed(&mut self) -> Result<()> {
		self.record_reset_impl(true)
	}

	fn commit_recorded_state(&mut self) {
		if let Some(seed) = self.pending_seed.take() {
			self.config.seed = seed;
		}
		if self.has_pending_full_reset {
			self.has_committed_state = true;
		}
		self.has_pending_full_reset = false;
	}

	fn rollback_recorded_state(&mut self) {
		self.pending_seed = None;
		self.has_pending_full_reset = false;
	}
}

fn validate_config(config: CartPoleConfig) -> Result<()> {
	let finite = [
		config.gravity,
		config.cart_mass,
		config.pole_mass,
		config.half_pole_length,
		config.force_magnitude,
		config.time_step,
		config.position_threshold,
		config.angle_threshold_radians,
	]
	.into_iter()
	.all(f32::is_finite);
	if !finite
		|| config.environments == 0
		|| config.max_episode_steps == 0
		|| config.environments > u32::MAX / 4
		|| config.cart_mass <= 0.0
		|| config.pole_mass <= 0.0
		|| config.half_pole_length <= 0.0
		|| config.force_magnitude <= 0.0
		|| config.time_step <= 0.0
		|| config.position_threshold <= 0.0
		|| config.angle_threshold_radians <= 0.0
	{
		return Err(Error::invalid_argument(
			"CartPole received an invalid environment configuration",
		));
	}
	Ok(())
}
